compiler/runtime/OMRRSSReport.cpp

```diff
@@ -0,0 +1,377 @@
+
+   TR_ASSERT_FATAL(offset >= 0, "Offset should be >= 0\n");
+
+   size_t tillPageEnd = _pageSize - ((uintptr_t)address % _pageSize);
+   size_t remainingSize = item->_size;
+
+   if (remainingSize > tillPageEnd)
+      item->_size = tillPageEnd;
+
+   addToListSorted(&_pageMap[offset], item);
+
+   remainingSize -= item->_size;
+   address += tillPageEnd;  // next address in case it's needed, points to the beginning of the next page
+
+   while (remainingSize > 0)
+      {
+      size_t  newSize = remainingSize;
+
+      if (newSize > _pageSize)
+         newSize = _pageSize;
+
+      OMR::RSSItem *newItem = new (TR::Compiler->persistentMemory()) OMR::RSSItem(item->_type, address, newSize, counters);
+      offset = (_grows == lowToHigh) ? ++offset : --offset;
+
+      TR_ASSERT_FATAL(offset >= 0, "Got negative offset %d for addr=%p size=%zu type=%s\n", offset,
+                      newItem->_addr, newItem->_size, OMR::RSSItem::itemNames[newItem->_type]);
+
+#ifdef DEBUG_RSS_REPORT
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS adding overflow item addr=%p size=%zu type=%s thread=%d method=%s",
+                                     newItem->_addr, newItem->_size,
+                                     OMR::RSSItem::itemNames[newItem->_type], threadId,
+                                     methodName ? methodName : "no method");
+#endif
+
+      addToListSorted(&_pageMap[offset], newItem);
+      address += _pageSize;
+      remainingSize -= newSize;
+      }
+   }
+
+void
+OMR::RSSRegion::addToListSorted(TR_PersistentList<OMR::RSSItem> *itemList, OMR::RSSItem *item)
+   {
+   ListElement<OMR::RSSItem> *prevElement = NULL;
+   ListElement<OMR::RSSItem> *newElement = NULL;
+   uint8_t *itemStart = item->_addr;
+   uint8_t *itemEnd = item->_addr + item->_size;
+   OMR::RSSItem *next = NULL;
+
+   ListIterator<OMR::RSSItem> it(itemList);
+
+   for (next = it.getFirst(); next; next = it.getNext())
+      {
+      uint8_t *nextEnd = next->_addr + next->_size;
+
+      if (itemStart < nextEnd)
+         {
+         newElement = itemList->addAfter(item, prevElement);
+         break;
+         }
+
+      prevElement = it.getCurrentElement();
+      }
+
+   if (!next)
+      newElement = itemList->addAfter(item, prevElement);
+
+   // Remove all subsequent overlaping items
+   // Due to the way we inserted the item above, all ovelaping items will be next
+   //
+   for (ListElement<OMR::RSSItem> * nextElement = newElement->getNextElement(); nextElement; nextElement = nextElement->getNextElement())
+      {
+      OMR::RSSItem *nextItem = nextElement->getData();
+      uint8_t *nextStart = nextItem->_addr;
+      uint8_t *nextEnd = nextStart + nextItem->_size;
+
+      if (itemStart >= nextStart || itemEnd > nextStart)
+         {
+         itemList->removeNext(newElement);
+         }
+      else
+         {
+         break;
+         }
+      }
+   }
+
+#ifdef LINUX
+typedef struct __attribute__ ((__packed__))
```

Comment:
The document does not mention `packed`. Can we assume all compilers support it? Otherwise, I am not sure how to check all of them. @0xdaryl

Reply:
Presumably, that attribute is required for efficiency rather than correctness.  If so, then I think guarding the typedef using `#ifdef` compiler macros just like in the example here should suffice:
https://github.com/eclipse/omr/blob/d58019eba46edbd5f2e1963e687a27c1dbaccac0/compiler/infra/String.hpp#L31
The `#else` part could just be a typedef without the attribute.  However, if the attribute is required for correctness for some reason then put an `#error` in the `#else` part instead.

Reply:
The attribute is needed for correctness. If some compiler does not support it I think it will give an error. Also, I noticed this attribute is being used in openj9 without any compiler check.

Reply:
OK, I don't think there are any non-Clang or non-gcc based compilers building and testing OMR.  The reference to xlC 12 in the supported compilers list is a bit dated, and we actually plan on revisiting that list at the next OMR architecture meeting.  Newer xlC's are based on Clang, I believe.  So I'll let this go.